ui/core/player.ts

```diff
@@ -1120,23 +1122,31 @@ export class Player<SpecType extends Spec> {
 		return this.computeStatsEP(stats);
 	}
 
+	computeUpgradeEP(equippedItem: EquippedItem, upgradeLevel: ItemLevelState, slot: ItemSlot): number {
+		const cacheKey = `${equippedItem.id}-${slot}-${equippedItem.randomSuffix?.id}-${upgradeLevel}`;
+		if (this.upgradeEPCache.has(cacheKey)) {
+			return this.upgradeEPCache.get(cacheKey)!;
+		}
+		const { item } = equippedItem;
+		const { item: upgradedItem } = equippedItem.withUpgrade(upgradeLevel).getWithDynamicStats();
+
+		let stats = new Stats(upgradedItem.stats);
+		stats = getWeaponStatsBySlot(item, slot);
```

Comment:
Looks like you're overwriting the first `stats` calculation here.

Reply:
Also, wouldn't you want to call `getWeaponStatsBySlot()` on the `upgradedItem` rather than the base one?